tests/common/test_json.py

```diff
@@ -348,3 +358,67 @@ def test_load_and_compare_all_impls() -> None:
         assert docs[idx] == docs[idx + 1]
         assert dump_s[idx] == dump_s[idx + 1]
         assert dump_b[idx] == dump_b[idx + 1]
+
+class X(BaseModel):
+    hello: str
+    url: AnyUrl
+
+def roundtrip(json_impl: SupportsJson) -> None:
+    x = X(url=AnyUrl("https://£££.com"), hello="world")
+    s = json_impl.dumps(x)
+    assert json_impl.loads(s) == {"url": "https://xn--9aaa.com/", "hello": "world"}
+
+@pytest.mark.parametrize("json_impl", _JSON_IMPL)
+def  test_serialize_custom_types_no_encoder(json_impl: SupportsJson) -> None:
+    with pytest.raises(TypeError, match="is not JSON serializable"):
+        roundtrip(json_impl)
+
+@pytest.mark.parametrize("json_impl", _JSON_IMPL)
+def  test_serialize_custom_types_noop_encoder(json_impl: SupportsJson) -> None:
+    calls = 0
+    def my_custom_encoder(obj: Any) -> Any:
+        nonlocal calls
+        calls += 1
+        if isinstance(obj, AnyUrl):
+            return None # returning None here will throw as if no encoder was set
+        return "non-none"
+
+    json_impl.set_custom_encoder(my_custom_encoder)
+    with pytest.raises(TypeError, match="is not JSON serializable"):
+        roundtrip(json_impl)
+    assert calls == 1
+
+@pytest.mark.parametrize("json_impl", _JSON_IMPL)
+def  test_serialize_custom_types_with_encoder(json_impl: SupportsJson) -> None:
+    calls = 0
+    def my_custom_encoder(obj: Any) -> Any:
+        nonlocal calls
+        calls += 1
+        if isinstance(obj, AnyUrl):
+            return str(obj)
+        return None
+
+    json_impl.set_custom_encoder(my_custom_encoder)
+    roundtrip(json_impl)
+    assert calls == 1
+
+@pytest.mark.parametrize("json_impl", _JSON_IMPL)
+def  test_serialize_custom_types_pua_string_only(json_impl: SupportsJson) -> None:
```

Comment:
this tests that the json implementation throws if the custom encoder returns non-`str` on a pua encode; see above.